### ISSUE TITLE
Switching between windows with Alt + Tab

### ISSUE THREAD

**Issue by petobens** [Hammerspoon/hammerspoon#856]
I'm trying to replace [HyperSwitch](https://bahoom.com/hyperswitch) functionality with Hammerspoon. Basically I want to be able to cycle between the windows in a screen (both active and minimized) using `Alt+Tab`. Ideally when cycling between windows there should be menu/display of the windows as thumbnails. Is this possible with Hammerspoon latest release? ( or does commit https://github.com/Hammerspoon/hammerspoon/commit/632b9a9053f4fe5e9d20318a720b53aa1d650aeb introduce this feature)? I've tried both expose and hints but they are not exactly what I'm looking for. Thanks!!

**Comment by heptal** [comment]
Yes, it's possible if you're willing to make it and wouldn't really be that difficult (at least conceptually/functionally) with what's available. Here's a related bit that bidirectionally cycles windows. It uses `window.filter`'s predicate function to constrain the set of cycleable windows to the current app and screen. You could go all out and use `hs.window:snapshot()`, `hs.drawing`, etc and make it as fancy as you want.
```lua
local currentWindowSet = {}
local windowCycler = nil

local wf = hs.window.filter.new(function(win)
    local fw = hs.window.focusedWindow()
    return (
      win:isStandard() and
      win:application() == fw:application() and
      win:screen() == fw:screen()
    )
  end)

local function makeTableCycler(t)
  local i = 1
  return function(d)
    local j = d and d < 0 and -2 or 0
    i = (i + j) % #t + 1
    local x = t[i]
    return x
  end
end

local function updateWindowCycler()
  if not hs.fnutils.contains(currentWindowSet, hs.window.focusedWindow()) then
    currentWindowSet = wf:getWindows()
    windowCycler = makeTableCycler(currentWindowSet)
  end
end

hs.hotkey.bind(hyper, "]", function()
    updateWindowCycler()
    windowCycler():focus()
  end)

hs.hotkey.bind(hyper, "[", function()
    updateWindowCycler()
    windowCycler(-1):focus()
  end)
```

**Comment by lowne** [comment]
Also check `hs.window.switcher`

**Comment by lowne** [comment]
Oops, it's not released yet, apologies. It'll be in the next release.

**Comment by petobens** [comment]
@lowne so `hs.window.switcher` essentially does exactly what I asked?

**Comment by lowne** [comment]
@petobens well, I can't be sure it does exactly what you need. By default (IIRC) it shows all visible windows on any screen/Space, but you can customise its windowfilter to include hidden windows, exclude other Spaces, change the display order etc. You can also change `.ui` to show/hide titles and snapshots. 

However 
- windowfilter doesn't (yet?) have a concept of "active screen" (i.e. the screen containing whatever the currently focused window is)
- windowswitcher always draws on the primary screen

So if by "cycle between windows in a screen" you mean e.g. having two screens, and when you're focused on the left one and hit alt-tab you want to see *on the left screen* all the windows that are on the left screen (and same for the right screen) then no, it can't currently do that.

I should probably add an active screen filter to windowfilter anyway (but I'm so short on time lately...), but the 2nd issue (in windowswitcher itself) isn't trivial.

**Comment by petobens** [comment]
@lowne thanks for your reply and explanation. Yes, when I said "cycle between windows in a screen" I meant exactly what you described: being able to use alt-tab to switch between the windows on the active screen. 
I''ll wait for the new release and test `hs.window.switcher` and close this issue afterwards (unless you want me to keep it open)

Once again, thanks!

**Comment by petobens** [comment]
@lowne now that there is a new release, can you please explain how to use `hs.window.switcher`? Especially how to bind it to `alt+tab` key combination. Thanks!

**Comment by petobens** [comment]
@lowne found the docs. Thanks anyway!

**Comment by heptal** [comment]
Please read them ;)

**Comment by petobens** [comment]
@lowne @heptal  how do I include minimized windows? Following the docs I tried with:
```
switcher = hs.window.switcher.new(hs.window.filter.new():setCurrentSpace(true):setDefaultFilter())
```

but minimized windows are not shown.

**Comment by petobens** [comment]
@lowne  can I add some code to my `init.lua` to always display the switcher on the active screen? or is it something you must add to the `hs.window.switcher` extension?

**Comment by asmagill** [comment]
@lowne, some additional notes:
  * you might want to change the default sizes, or at least document more clearly that people using a laptop without an external monitor need to adjust the defaults -- the defaults put the switcher bar off screen on my MacBook Air (1440x900, both Dock and Menubar showing, so usable is slightly less)
  * previousWindow should be a function in the docs, not a method

I'll check it out on a multi-monitor setup later

**Comment by heptal** [comment]
@petobens try reading the hs.window.filter documentation. The Google group might be more suitable for general usage questions that don't relate to technical issues with Hammerspoon.

**Comment by petobens** [comment]
@heptal I read (at least a bit) of the `hs.window.filter` documentation, however I asked about using:

>switcher = hs.window.switcher.new(hs.window.filter.new():setCurrentSpace(true):setDefaultFilter())

because according to [`hs.window.switcher` docs](http://www.hammerspoon.org/docs/hs.window.switcher.html) it should show minimized windows but (at least for me) it doesn't.

**Comment by heptal** [comment]
The windowfilter determines the set of windows switched...

hs.window.switcher is brand new so there may be some inaccuracies but the windowfilter docs say that the default filter excludes invisible (incl minimized) windows. It's very easy to just try different filters.

**Comment by lowne** [comment]
@petobens yep, the documentation is wrong :) - the correct syntax is `:setDefaultFilter{}`, fixed now, thanks! (Also: minimised windows were not unminimised upon selection, that's also fixed now). For the "active screen" part, see #868

@asmagill thanks; I've reduced the default thumbnail size to 384, it should fit now - if not, feel free to further tweak the numbers in `ui`. Fixed docstring for previousWindow as well.
